File: LC/python/Word Search.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        def dfs(i,j,index):
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or word[index] != board[i][j]:
                return False
            if index == len(word) -1:
                return True
            board[i][j] = '#'
            res = dfs(i+1,j,index+1) or dfs(i-1,j,index+1) or dfs(i,j+1,index+1) or dfs(i,j-1,index+1)
            board[i][j] = word[index]
            return res
            
        res = False
        for m in range(len(board)):
            for n in range(len(board[0])):
                res = res or dfs(m,n,0)
        return res
        
        
        

File: LC/java/Word Search.java
public class Solution {
    public boolean exist(char[][] board, String word) {
        boolean res = false;
        
        for(int m = 0; m < board.length; m++){
            for(int n = 0; n<board[0].length; n++){
                res = res || dfs(m,n,0,board,word);
            }
        }
        return res;
    }

    public boolean dfs(int i, int j, int index, char[][] board, String word){
        if(i<0 || j<0 || i>=board.length || j>=board[0].length || board[i][j]!=word.charAt(index)){
            return false;
        }
        if(index == word.length()-1){
            return true;
        }
        board[i][j] = '#';
        
        boolean res = dfs(i+1,j,index+1,board,word) || dfs(i,j+1,index+1,board,word) || dfs(i-1,j,index+1,board,word) || dfs(i,j-1,index+1,board,word);
        
        board[i][j] = word.charAt(index);
        
        return res;
    }
}
